cartofacade/cgo_api.go

```diff
@@ -0,0 +1,401 @@
+	defer C.free_alloc_viam_carto_lib(pVcl)
+
+	ppVcl := (**C.viam_carto_lib)(unsafe.Pointer(&pVcl))
+	/*
+		Question: what is going on here? are we not getting correct amount of indirection?
+					fmt.Printf("%#v\n", pVcl) = &cartofacade._Ctype_struct_viam_carto_lib{minloglevel:0, verbose:0}
+					fmt.Println(&pVcl) = 0x14000128078
+					fmt.Println(goPtr) = 0x600001ee8000
+					// ppVcl := (**C.viam_carto_lib)(goPtr)
+					fmt.Println(ppVcl) = 0x600001ee8000
+	*/
+
+	status := C.viam_carto_lib_init(ppVcl, C.int(miniloglevel), C.int(verbose))
+	if err := getErrorFromStatusCode(status); err != nil {
+		return CartoLib{}, err
+	}
+
+	vcl := CartoLib{value: (*ppVcl)}
+
+	return vcl, nil
+}
+
+// TerminateCartoLib calls viam_carto_lib_terminate to clean up memory for viam carto lib.
+func (vcl *CartoLib) TerminateCartoLib() error {
+	status := C.viam_carto_lib_terminate((**C.viam_carto_lib)(unsafe.Pointer(vcl)))
+	if err := getErrorFromStatusCode(status); err != nil {
+		return err
+	}
+	return nil
+}
+
+// NewCarto calls viam_carto_init and returns a pointer to a viam carto object.
+func NewCarto(cfg CartoConfig, acfg CartoAlgoConfig, vcl CartoLib) (Carto, error) {
+	pVc := C.alloc_viam_carto()
+	defer C.free_alloc_viam_carto(pVc)
+
+	ppVc := (**C.viam_carto)(unsafe.Pointer(&pVc))
+
+	vcc := getConfig(cfg)
+	defer C.free_bstring_array(vcc.sensors)
+
+	vcac := getAlgoConfig(acfg)
+
+	status := C.viam_carto_init(ppVc, vcl.value, vcc, vcac)
+	if err := getErrorFromStatusCode(status); err != nil {
+		return Carto{}, err
+	}
+
+	return Carto{value: (*ppVc)}, nil
+}
+
+// Start is a wrapper for viam_carto_start
+func (vc *Carto) Start(ctx context.Context) error {
+	status := C.viam_carto_start(vc.value)
+
+	if err := getErrorFromStatusCode(status); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Stop is a wrapper for viam_carto_stop
+func (vc *Carto) Stop(ctx context.Context) error {
+	status := C.viam_carto_stop(vc.value)
+
+	if err := getErrorFromStatusCode(status); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// TerminateCarto calls viam_carto_terminate to clean up memory for viam carto
+func (vc *Carto) TerminateCarto(ctx context.Context) error {
+	status := C.viam_carto_terminate(&vc.value)
+
+	if err := getErrorFromStatusCode(status); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// AddSensorReading is a wrapper for viam_carto_add_sensor_reading
+func (vc *Carto) AddSensorReading(ctx context.Context, readings []byte, timestamp time.Time) error {
+	sensorReading := C.viam_carto_sensor_reading{}
+	sensorReading.sensor_reading = C.blk2bstr(unsafe.Pointer(&readings[0]), C.int(len(readings)))
+
+	sensorReading.sensor_reading_time_unix_micro = C.ulonglong(timestamp.UnixMicro())
+
+	resp := AddSensorReading{value: sensorReading}
+
+	status := C.viam_carto_add_sensor_reading(vc.value, &resp.value)
+
+	if err := getErrorFromStatusCode(status); err != nil {
+		return err
+	}
+
+	status = C.viam_carto_add_sensor_reading_destroy(&resp.value)
```

Comment:
This needs to be deferred after `viam_carto_add_sensor_reading` so that we do it even if we get an error from adding a sensor reading